### PR TITLE
Canon CR2: white balance

### PR DESCRIPTION
**Pull request by LebedevRI**
This was developed against DT copy of rawspeed, that already contains all those yet unmerged changes by @pedrocr 
Specifically, it needs macrosed like get2BE() that are part of PR #73 
Until that PR is merged, this one _should_ not be merged.

With this, RawSpeed can read WB info from ALL CR2 files (at least at this moment, all those cr2 sample files from rawsamples.ch)

### CODE REVIEW

File: RawSpeed/Cr2Decoder.cpp [hunk 1]
```diff
@@ -190,6 +191,104 @@ void Cr2Decoder::decodeMetaDataInternal(CameraMetaData *meta) {
...
+      case 1346: /* Canon EOS 60D, Canon EOS 1100D */
+        offset += 126;
+      break;
+      case 5120:
+        /* Canon PowerShot G10, Canon PowerShot G11, Canon PowerShot G12,
+         * Canon PowerShot G1X, Canon PowerShot S120, Canon PowerShot SX1 IS */
+        offset += 142;
+      break;
+      default:
+        std::cerr << "CR2 Decoder: CanonColorData has unsupported count of values: %d" << color_data->count << std::endl;
```
Comment (klauspost) [inline RawSpeed/Cr2Decoder.cpp]: Please don't print to stderr. Use writeLog(DEBUG_PRIO_INFO, "CR2 Decoder: CanonColorData has unsupported count of values: %d", color_data->count);

File: RawSpeed/Cr2Decoder.cpp [hunk 2]
```diff
@@ -190,6 +191,104 @@ void Cr2Decoder::decodeMetaDataInternal(CameraMetaData *meta) {
...
     iso = mRootIFD->getEntryRecursive(ISOSPEEDRATINGS)->getInt();
 
+  // Fetch the white balance
+  if (mRootIFD->hasEntryRecursive(CANONCOLORDATA)) {
+    TiffEntry *color_data = mRootIFD->getEntryRecursive(CANONCOLORDATA);
+
+    // this entry is a big table, and different cameras store used WB in
+    // different parts, so find the offset
+    int offset = 0;
+    switch (color_data->count)
```
Comment (klauspost) [inline RawSpeed/Cr2Decoder.cpp]: Please use this value as a _default_ value, add a hint that allow us to override the value like this

`<Hint name="wb_offset" value="72"/>`

That way, we can add support for new cameras without modifying code.

File: RawSpeed/Cr2Decoder.cpp [hunk 3]
```diff
@@ -190,6 +191,104 @@ void Cr2Decoder::decodeMetaDataInternal(CameraMetaData *meta) {
...
+     * as TIFF_UNDEFINED, while they still write normal TIFF_SHORT data there
+     */
+    if (color_data->type == TIFF_SHORT || color_data->count == 5120) {
+      const ushort16* data = color_data->getShortArray();
+
+      // RGGB !
+      float cam_mul[4];
+      for(int c = 0; c < 4; c++)
+      {
+        cam_mul[c] = (float) data[offset/2 + c];
```
Comment (klauspost) [inline RawSpeed/Cr2Decoder.cpp:221]: data[offset/2 + c] needs to be bound-schecked against the tiff entry, otherwise we can read junk/crash.

**Comment by klauspost** [comment]
Looks good. I had a quick look, and added a few notes. #73 should be merged soon, so this comes at a good time!

**Comment by pedrocr** [comment]
I've fixed the 1D/1Ds support including WB for them so after you merge this one I'll update the PR

**Comment by klauspost** [comment]
Cool - just fixing a leak in CRW. After I have verified it, I will look at this.

**Pull request merged**